### PR TITLE
SPC toolkit python package

### LINKED ISSUE #55: Add python build interface to run python scripts/notebooks
Include a build interface (`pyproject.toml`) for reproducible running of python scripts and notebooks.

- [x] Add `pyproject.toml` with poetry and confirm package requirements
- [x] Check all scripts and notebooks can be run upon installation
- [x] Add install instructions

### PR DESCRIPTION
**Pull request by sgreenbury**
Closes #55.

Remaining tasks:
- [x] Refactor to single class for reading protobuf or parquet
- [x] Complete support for pandas backend:
  - [x] Reader
  - [x] Merging
- [x] Refine merging methods/implement dataframe builder
- [x] Update dependencies and create `dev` and `examples` groups
- [x] Update notebook examples
- [x] Add build and running tests to actions
- [x] Update README with different install options including pip
- [x] Check serialized population for flat output matches single protobuf structure
- [x] Test examples
- [x] Update test data
- [x] Remove `#[derive(Serialize)` on types no longer requiring serialize
- [x] Add test for determinism in parquet and JSON 
- [x] Fix unnest exception when column exists in dataframe after unnesting (e.g. `nssec8`)
- [x] Implement reading from remote gzip
- [x] Re-export `Reader` and `Builder` classes
- [x] Update CLI with `--output-formats Vec<OutputFormat>`
- [x] Update pages docs

### CODE REVIEW

**Comment by dabreegster** [comment]
Hi Sam, @Hussein-Mahfouz would like to use this; need a beta tester?

**Comment by sgreenbury** [comment]
That would be great, thanks! @Hussein-Mahfouz the [`python/README.md`](https://github.com/alan-turing-institute/uatk-spc/blob/55-output-formats-python/python/README.md) is the best place to start but happy to discuss if unclear at all. Just to flag the API will change while this PR is in draft and the remaining features at the top are added.

Any comments on the usability/missing features would be very helpful!

**Comment by sgreenbury** [comment]
Hi @Hussein-Mahfouz, thanks for trying this out! Currently the instructions in the [README](https://github.com/alan-turing-institute/uatk-spc/blob/55-output-formats-python/python/README.md) assume you are on the branch associated with this PR (`55-output-formats-python`) and are in the the `python/` directory (where the README and [`pyproject.toml`](https://github.com/alan-turing-institute/uatk-spc/blob/55-output-formats-python/python/pyproject.toml) are), apologies that wasn't clear. From the repo root if you try:
```bash
git checkout 55-output-formats-python
cd python
```
then `poetry install` should work as you will now have the `pyproject.toml` and package.

**Comment by sgreenbury** [comment]
Removed upper limit on required python version ([ad1d826](https://github.com/alan-turing-institute/uatk-spc/pull/67/commits/ad1d8267e563348bd9634f5773c64bd98e76bd96), see https://github.com/Urban-Analytics-Technology-Platform/acbm/issues/10).

File: python/README.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+- [Python](https://www.python.org/)
+- [Poetry (optional)](https://python-poetry.org/)
+
+## Quickstart
+### Generating outputs
+From the repo root, checkout the branch:
+```
+git checkout 55-output-formats-python
+```
+and then run SPC on a region with `--flat-output` (e.g. Rutland is shown below):
```
Comment (sgreenbury) [inline python/README.md:16]:
```suggestion
From the repo root, run SPC on a region with `--flat-output` (e.g. Rutland is shown below):
```

File: python/README.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+```
+cargo run --release -- --rng-seed 0 --flat-output config/England/rutland.txt
+```
+
+### Install
+The package can be installed with `pip` from git with:
+```bash
+pip install 'git+https://github.com/alan-turing-institute/uatk-spc.git@55-output-formats-python#subdirectory=python'
```
Comment (sgreenbury) [inline python/README.md:24]:
```suggestion
pip install 'git+https://github.com/alan-turing-institute/uatk-spc.git#subdirectory=python'
```

File: python/README.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+```
+or with Poetry:
+```bash
+poetry add 'git+https://github.com/alan-turing-institute/uatk-spc.git@55-output-formats-python#subdirectory=python'
```
Comment (sgreenbury) [inline python/README.md:28]:
```suggestion
poetry add 'git+https://github.com/alan-turing-institute/uatk-spc.git#subdirectory=python'
```

File: python/README.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+```
+
+#### Extras
+To include extra dependencies `dev` and `examples` for running [tests](./tests/) and [examples](./examples/):
+```
+pip install 'uatk-spc[dev,examples] @ git+https://github.com/alan-turing-institute/uatk-spc.git#subdirectory=python'
+```
+or with Poetry:
+```bash
+poetry add 'git+https://github.com/alan-turing-institute/uatk-spc.git@55-output-formats-python#subdirectory=python'
```
Comment (sgreenbury) [inline python/README.md:38]:
```suggestion
poetry add 'git+https://github.com/alan-turing-institute/uatk-spc.git#subdirectory=python'
```

File: web/components/Diaries.svelte
Comment (sgreenbury) [inline web/components/Diaries.svelte:1]: Move these changes to another PR

**Review by dabreegster — COMMENTED** [review]
I skimmed through the Rust pieces, and everything looks good. Clever use of serde and converters to keep boilerplate low for handling `TiVec` and the prost-generated types :)

**Comment by sgreenbury** [comment]
Thanks @dabreegster!
